utils/pattern/src/interpolator/mod.rs

```diff
@@ -0,0 +1,322 @@
+/// `Option<Result<Item>>`).
+///
+/// Since the interpolator chains on top of the [`Parser`] it inherits the same fallible
+/// iterator API and behavior.
+///
+/// [`Item`]: std::iter::Iterator::Item
+/// [`HashMap`]: std::collections::HashMap
+pub struct Interpolator<'p, R, K, E>
+where
+    R: ReplacementProvider<K, E>,
+{
+    pattern: Parser<'p>,
+    replacements: R,
+    current_replacement: Option<R::Iter>,
+}
+
+impl<'p, R, K, E> Interpolator<'p, R, K, E>
+where
+    R: ReplacementProvider<K, E>,
+{
+    /// Creates a new `Interpolator`.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, Interpolator};
+    ///
+    /// enum Element {
+    ///     Literal(String),
+    ///     Token,
+    /// }
+    ///
+    /// let mut parser = Parser::new("{0}, {1}");
+    /// let mut interpolator = Interpolator::new(parser, vec![
+    ///     vec![
+    ///         Element::Token
+    ///     ]
+    /// ]);
+    /// ```
+    pub fn new(pattern: Parser<'p>, replacements: R) -> Self {
+        Self {
+            pattern,
+            replacements,
+            current_replacement: None,
+        }
+    }
+
+    /// An iterator method that advances the iterator and returns the result of an attempt to
+    /// interpolate parser and replacement provider tokens.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, Interpolator};
+    ///
+    /// #[derive(Debug, PartialEq)]
+    /// enum Element {
+    ///     Literal(String),
+    ///     TokenOne,
+    ///     TokenTwo,
+    /// }
+    ///
+    /// impl From<&str> for Element {
+    ///     fn from(input: &str) -> Self {
+    ///        Self::Literal(input.to_string())
+    ///     }
+    /// }
+    ///
+    /// let mut parser = Parser::new("{0}, {1}");
+    /// let mut interpolator = Interpolator::new(parser, vec![
+    ///     vec![
+    ///         Element::TokenOne
+    ///     ],
+    ///     vec![
+    ///         Element::TokenTwo
+    ///     ]
+    /// ]);
+    ///
+    ///
+    /// // A call to try_next() returns the next value…
+    /// assert_eq!(Ok(Some(Element::TokenOne)), interpolator.try_next());
+    /// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+    /// assert_eq!(Ok(Some(Element::TokenTwo)), interpolator.try_next());
+    ///
+    /// // … and then None once it's over.
+    /// assert_eq!(Ok(None), interpolator.try_next());
+    /// ```
+    pub fn try_next(&mut self) -> Result<E, K>
+    where
+        E: From<&'p str>,
+        K: FromStr + std::fmt::Display,
+        K::Err: std::fmt::Debug,
+    {
+        loop {
+            if let Some(ref mut replacement) = &mut self.current_replacement {
+                if let Some(v) = replacement.next() {
+                    return Ok(Some(v));
+                } else {
+                    self.current_replacement = None;
+                }
+            }
+            match self
+                .pattern
+                .try_next()
```

Comment:
Issue: As written, the interpolator parses the pattern freshly each time.  It would be more efficient, elegant, and smaller code if you included a Pattern type with everything pre-parsed.  This type could be used in data structs, etc.

Reply:
The interpolator doesn't parse, the interpolator combines the pattern iterator with replacement iterators.

My intention is to allow feeding the interpolator a pre-parsed pattern just like you're asking. In fact, for most of life of the test branch that's what it was, the input to `Interpolator` was an `Iterator`, but I just realized that I lost it when I switched my model to pseudo fallible-iterator.
I'll think about how to regain it, since that's definitely the goal here!

One idea that would solve it with minimal overhead would be to introduce (temporarily, until Rust adds fallible iterators) a custom `FallibleIterator` trait that would be implemented on `Parser` and we'd implement it on `Vec<PatternToken>`.

Reply:
I added `PatternIterator` trait and implemented it for `Parser` and added `IntoIterVec` with `From<Vec>` and also implemented it on it.

The result is that you can now pass either a parser to the `Interpolator` or a `IntoIterVec` constructed from a `Vec`.

This is still a bit clumsy, but in practice it should lead to a readable code and easy migration once Rust starts handling fallible iterators.